lib/wibox/widget/separator.lua

```diff
@@ -0,0 +1,188 @@
+--
+-- * `wibox.layout.fixed`
+-- * `wibox.layout.flex`
+-- * `wibox.layout.ratio`
+--
+-- When used with these layouts, it is also possible to provide custom clipping
+-- functions. This is useful when the layout has overlapping widgets (negative
+-- spacing).
+--
+--@DOC_wibox_widget_defaults_separator_EXAMPLE@
+--
+-- @author Emmanuel Lepage Vallee &lt;elv1313@gmail.com&gt;
+-- @copyright 2014, 2017 Emmanuel Lepage Vallee
+-- @classmod wibox.widget.separator
+---------------------------------------------------------------------------
+local beautiful = require( "beautiful"         )
+local base      = require( "wibox.widget.base" )
+local color     = require( "gears.color"       )
+local gtable    = require( "gears.table"       )
+
+local separator = {}
+
+--- The separator's orientation.
+--
+-- Valid values are:
+--
+-- * *vertical*: From top to bottom (default)
+-- * *horizontal*: From left to right
+--
+--@DOC_wibox_widget_separator_orientation_EXAMPLE@
+--
+-- @property orientation
+-- @param string
+
+--- The separator's thickness.
+--
+-- This is used by the default line separator, but ignored when a shape is used.
+--
+-- @property thickness
+-- @param number
+
+--- The separator's reflection.
+--
+-- When a shape or a custom drawing function is used, the default shape assumes
+-- this to go from left to right.
+-- When this property is set to anything else, the shape will be rotated on one
+-- axis or another.
+--
+-- This is the same as `wibox.container.mirror`, but need to exist on the widget
+-- level because it affects other properties such as the layout clipping.
+--
+-- @property reflection
+-- @tparam table reflection
+-- @tparam[opt=false] boolean reflection.horizontal
+-- @tparam[opt=false] boolean reflection.vertical
+
+--- The separator's shape.
+--
+--@DOC_wibox_widget_separator_shape_EXAMPLE@
+--
+-- @property shape
+-- @tparam function shape A valid shape function
+-- @see gears.shape
+
+--- The relative percentage covered by the bar.
+-- @property span_ratio
+-- @tparam[opt=1] number A number between 0 and 1.
+
+--- The separator's color.
+-- @property color
+-- @param string
+-- @see gears.color
+
+--- The separator's border color.
+--
+--@DOC_wibox_widget_separator_border_color_EXAMPLE@
+--
+-- @property border_color
+-- @param string
+-- @see gears.color
+
+--- The separator's border width.
+-- @property border_width
+-- @param number
+
+local function draw_shape(self, _, cr, width, height, shape)
+    local bw = self._private.border_width or 0
+
+    cr:translate(bw/2, bw/2)
+
+    shape(cr, width-bw, height-bw)
+
+    if bw == 0 then
+        cr:fill()
+    elseif self._private.border_color then
+        cr:fill_preserve()
+        cr:set_source(color(self._private.border_color))
+        cr:set_line_width(bw)
+        cr:stroke()
+    end
+end
+
+local function draw_line(self, _, cr, width, height)
+    local thickness = self._private.thickness or beautiful.separator_thickness or 1
```

Comment:
`separator_thickness` is undocumented.

Also, I'm a bit confused about `separator_thickness` and `span_ratio`. One is a percentage of the "width" (in a horizontal line) while the other is a fixed size in pixels. I can see why the width should get a percentage, but still this feels inconsistent to me.

Reply:
> One is a percentage of the "width" (in a horizontal line) while the other is a fixed size in pixels

I don't have much to say beside "it scales better when you resize the wibar", but it's a tiny corner case. When you have a larger wibar/notification, you don't want to have a thicker separator. But keeping the same number of pixel for the length just looks wrong (to me).